pkg/util/kubernetes/crd.go

```diff
@@ -16,23 +16,64 @@ package kubernetes
 
 import (
 	"encoding/json"
-	"errors"
 	"fmt"
+	"reflect"
 	"time"
 
-	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
-	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"github.com/sirupsen/logrus"
+	extsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	extsclientset "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	watchapi "k8s.io/apimachinery/pkg/watch"
 	"k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/watch"
 
 	"github.com/nats-io/nats-operator/pkg/apis/nats/v1alpha2"
-	"github.com/nats-io/nats-operator/pkg/debug/local"
-	"github.com/nats-io/nats-operator/pkg/util/retryutil"
+	contextutil "github.com/nats-io/nats-operator/pkg/util/context"
+)
+
+const (
+	// waitCRDReadyTimeout is the maximum period of time we wait for each CRD to be ready (i.e. established).
+	waitCRDReadyTimeout = 30 * time.Second
 )
 
 var (
-	ErrCRDAlreadyExists = errors.New("crd already exists")
+	// crds contains all the custom resource definitions that nats-operator registers upon starting.
```

Comment:
Eventually we could look into having the operator not be responsible of creating these CRDs, would make it possible to narrow down some of the required policy for it, for now I added them to the manifests as well so that would be present already when deploying the operator.
https://github.com/nats-io/nats-operator/issues/81

Reply:
This doesn't make much sense to me as it limits what one can do with and to CRDs, like migration. Also, relying on the user seems to be exactly what we're avoiding by automating with an operator ;)

Reply:
The end-goal is for CRD multi-versioning, btw.

Reply:
Right ok can be addressed when setting up for multi version CRD from k8s v1.11